cli/azd/cmd/cobra_builder.go

```diff
@@ -0,0 +1,221 @@
+	if descriptor.Parent() == nil {
+		if err := cb.bindCommand(cmd, descriptor); err != nil {
+			return nil, err
+		}
+	}
+
+	cb.configureActionResolver(cmd, descriptor)
+
+	return cmd, nil
+}
+
+// Configures the cobra command 'RunE' function to running the composed middleware and action for the
+// current action descriptor
+func (cb *CobraBuilder) configureActionResolver(cmd *cobra.Command, descriptor *actions.ActionDescriptor) {
+	// Only bind command to action if an action resolver had been defined
+	// and when a RunE hasn't already been set
+	if descriptor.Options.ActionResolver == nil || cmd.RunE != nil {
+		return
+	}
+
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		ctx := cmd.Context()
+		ctx = tools.WithInstalledCheckCache(ctx)
+
+		// Registers the following to enable injection into actions that require them
+		ioc.RegisterInstance(ioc.Global, cb.runner)
+		ioc.RegisterInstance(ioc.Global, ctx)
+		ioc.RegisterInstance(ioc.Global, cmd)
+		ioc.RegisterInstance(ioc.Global, args)
+
+		if err := cb.registerMiddleware(descriptor); err != nil {
+			return err
+		}
+
+		actionName := createActionName(cmd)
+		var action actions.Action
+		if err := cb.container.ResolveNamed(actionName, &action); err != nil {
+			return fmt.Errorf(
+				//nolint:lll
+				"failed resolving action '%s'. Ensure the ActionResolver is a valid go function that returns an `actions.Action` interface, %w",
+				actionName,
+				err,
+			)
+		}
+
+		runOptions := &middleware.Options{
+			Name:    cmd.CommandPath(),
+			Aliases: cmd.Aliases,
+		}
+
+		// Run the middleware chain with action
+		log.Printf("Resolved action '%s'\n", actionName)
+		_, err := cb.runner.RunAction(ctx, runOptions, action)
+
+		return err
+	}
+}
+
+// Binds the intersection of cobra command options and action descriptor options
+func (cb *CobraBuilder) bindCommand(cmd *cobra.Command, descriptor *actions.ActionDescriptor) error {
+	actionName := createActionName(cmd)
+
+	// Automatically adds a consistent help flag
+	cmd.Flags().BoolP("help", "h", false, fmt.Sprintf("Gets help for %s.", descriptor.Name))
+
+	// Consistently registers output formats for the descriptor
+	if len(descriptor.Options.OutputFormats) > 0 {
+		output.AddOutputParam(cmd, descriptor.Options.OutputFormats, descriptor.Options.DefaultFormat)
+	}
+
+	// Create, register and bind flags when required
+	if descriptor.Options.FlagsResolver != nil {
+		log.Printf("registering flags for action '%s'\n", actionName)
+		ioc.RegisterInstance(cb.container, cmd)
+
+		// The flags resolver is constructed and bound to the cobra command via dependency injection
+		// This allows flags to be options and support any set of required dependencies
+		if err := cb.container.RegisterSingletonAndInvoke(descriptor.Options.FlagsResolver); err != nil {
+			return fmt.Errorf(
+				"failed registering FlagsResolver for action '%s'. Ensure the resolver is a valid go function. %w",
```

Comment:
Since we also invoke the constructor here, which can include an `err`:

```suggestion
				"failed registering FlagsResolver for action '%s'. Ensure the resolver is a valid go function and resolves without err. %w",
```